File: 1364-tuple-with-same-product/tuple-with-same-product.cpp
class Solution {
public:
    int tupleSameProduct(vector<int>& nums) {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);
        int n = nums.size(), res = 0;
        unordered_map<int,int> mp;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                mp[nums[i] * nums[j]]++;
            }
        }
        for (auto &[k, v] : mp) {
            res += v * (v - 1);
        }
        return res * 4;
    }
};

File: 1364-tuple-with-same-product/tuple-with-same-product.rs
use std::collections::HashMap;

impl Solution {
    pub fn tuple_same_product(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut res = 0;
        let mut mp: HashMap<i32, i32> = HashMap::new();

        for i in 0..n {
            for j in (i + 1)..n {
                *mp.entry(nums[i] * nums[j]).or_insert(0) += 1;
            }
        }

        for (_, v) in mp.iter() {
            res += v * (v - 1);
        }

        res * 4
    }
}